### ISSUE TITLE
IOS only - player will not play after reaching end of media and emitting ProcessingState.completed

### ISSUE THREAD

**Issue by fuzing** [ryanheise/just_audio#152]
**Which API doesn't behave as documented, and how does it misbehave?**
Note - this issue occurs on IOS only.
After the player emits ProcessingState.completed, when seeking backwards (anywhere) and attempting to call .play(), the player will not play the audio.   Calling any combination of .pause(), .play(), .seek() fails to restart the player.

**Minimal reproduction project**
Happy to work on this and provide if necessary.

**To Reproduce (i.e. user steps, not code)**

**Error messages**
No errors being emitted.

**Expected behavior**
Would expect that after reaching the completed state, seeking, playing would undo this state (it may be doing this) and allowing playback from earlier in the audio.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**
 - iPad
 - OS: IOS 12.4.8

**Flutter SDK version**
1.21

insert output of "flutter doctor" here
[✓] Flutter (Channel master, 1.21.0-10.0.pre.167, on Mac OS X 10.15.4 19E266, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
[!] Android Studio (version 4.0)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
[✓] VS Code (version 1.48.0)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.

I'm using VS, not Android Studio, so can ignore issue!

**Additional context**
N/A

**Comment by ryanheise** [comment]
Hi @fuzing , maybe this should be reproducible in the example if you skip to the last track, seek to just before the end, and let it play out to completion, and then click the seek bar to seek back. In my case, though, it correctly resumes playback. Since it doesn't happen in the example, then yes a minimal reproduction project will be needed.

**Comment by fuzing** [comment]
Agreed.... thank you for that feedback.  I'll download the sample to my IOS device and confirm that it works with my IOS version and hardware combo.... then I'll take a look at the sample code vs. mine for clues.  Failing this, I'll assemble a minimum test app.

**Comment by fuzing** [comment]
I'm just testing also - will have feedback in about 10 minutes

**Comment by fuzing** [comment]
Looks good to me - I'll wait an hour or two and then close - or Ryan, be my guest.

Thanks for your work on this.    Very quick question - was this an IOS only fix, or did you make mods to code that might impact both IOS and Android? (I ask insofar as I have a similar [albeit different] issue with Android not playing at all, but only after an extended period of time - i.e. I leave the device for a day or so with the background task up, but no foreground - then come back and nothing plays.

Let me know.  I'm in the process of trying to replicate in a consistent manner (maybe its gone now?).

**Comment by ryanheise** [comment]
This was an iOS-only change, so this sounds like it would be a different issue.

One thing I should mention about the audio-session branch (which just just been merged) is that it has been tested only with the audio-session branch of audio_service. I am not sure whether audio focus handling would work if only just_audio were updated and audio_service were not. The reason is that this change moves audio focus management out of audio_service. But if you use the new just_audio and the old audio_service, you'll end up with audio focus handling code in two places.

Thus, I would like to release the two plugins roughly simultaneously, although I would feel more confident about doing that after hearing from people using the new audio_service (audio-session branch).

**Comment by fuzing** [comment]
Hi Ryan - For my testing (quick and dirty) I didn't bring the two plugins/repos into alignment........ don't think I'll be able to get to that until tomorrow, but will definitely do that and circle back around.  Thanks for the heads-up.

**Comment by fuzing** [comment]
I'll need to pin down exactly what's going on for the Android problem so as not to waste anyone's time.   Before attempting to draft a synopsis I'll make sure I'm on the latest versions (or branches/refs if that would be preferable).......

**Comment by fuzing** [comment]
Ryan - to get the right combination........
Is audio_service -> audio-session branch now merged into master (I can't find an audio-session tag/branch)

**Comment by ryanheise** [comment]
Yes it has now been merged into master. I realised I didn't actually update the example completely, but the plugin itself should be fine on master. The example should work, although it will by default set up the audio session for music playback, and I had intended to explicitly set it to speech for this example (I am in the process updating the example further.)

**Comment by fuzing** [comment]
Great - just want to make sure I'm in sync.

**Comment by ryanheise** [comment]
The simplest way to migrate is to remove any audio interruption handlers from your background audio task (if you have any), and then pass `handleInterruptions: true` into just_audio's constructor to let it handle it.

Everything else is under the hood unless you want to change the default behaviour of the underlying audio session. An audio session controls how an app takes ownership of audio on the device. Depending on the nature of your app, you may want to pause or duck if another app plays audio, or you may alternatively want to cause other apps to duck or pause. The defaults are reasonable for a music app, but if you require special behaviour, you can then use the audio_session package to directly declare your app's audio requirements and how you would like to interact with other audio apps.

The one thing I haven't decided on yet is whether to make `handleInterruptions: true` the default behaviour for just_audio. The primary reason for defaulting to `false` is because previous versions of just_audio didn't handle interruptions. But if I think about it, probably it would be reasonable to enable it by default, and let people pass in `false` if they have more specialised behaviour that they want to implement themselves.

**Comment by fuzing** [comment]
I'd agree - defaulting to handleInterruptions: true    sounds like a
reasonable default option..... flattens the learning curve and makes
adoption much easier.   I can't speak to everyone else in terms of where
they're at, but I'm handling those interruptions in "service" and doing
what I believe are the appropriate gymnastics with just_audio...... easy
for me to change, and I'd be happy to hand that control over.

That being said, others may be surprised/caught off guard if the default is
to wrestle away control....... hence if it were me, I'd default to false
for this major.minor combination and flip it for the next major (with the
appropriate release note) (just my 2 cents).

Thanks for all your help with this.   Know that I (and others) really
appreciate your efforts.

**Comment by ryanheise** [comment]
That was my original thinking, and I was originally planning a phased approach with the next release of just_audio being perfectly backward compatible with the old version by default. However, just defaulting `handleInterruptions` to false doesn't alone make it backward compatible, although I had initially hoped it would. A truly backward compatible version would look quite ugly because it would have to emulate the fact that the old version of just_audio handled things inconsistently on Android and iOS. That is, the old version did audio session management for iOS but not Android.

But theoretically it would be possible if I had a constructor that looks like this:

```dart
AudioPlayer({handleInterruptions: false, backwardsCompatible: true});
```

And internally do some horrible looking code that people won't see.

Then in audio_service I could add back in the audio focus callbacks and deprecate them.

**Comment by fuzing** [comment]
Ryan,

Do you want me to close this?  I think this is fixed.

Ugggh...... that's the problem with backward compatibility.    At some point you just have to increment the major and move on...... screams willl fade eventually (earplugs are good) and the codebase and plugin will be cleaner/better for it, and you'll sleep better.  This project has moved quickly and as long as the left behind version can be minor patched I think most people get it.  You're one person, not a team of 20.  Well done.

**Comment by fuzing** [comment]
BTW, my android issue was as a result of an exception being thrown due to my code in onClose() in the derived background task class..... resulting in super.onClose() not being called - my fault......

**Comment by ryanheise** [comment]
Which onClose is this? If it is audio_session, I should mention that ideally I would like to get rid of this method and have the cleanup handled automatically by the plugin.

**Comment by ryanheise** [comment]
Ah, I see you mean in audio_service.

**Comment by fuzing** [comment]
Yes - in audio_service - sorry.   I'm calling directly on Android
swipe-away.

**Comment by ryanheise** [comment]
OK, I'll close this issue then, and thanks for the perspective on backwards compatibility. I'll ponder it further, and a proper release will probably happen in a day or two.
